src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -13,74 +20,101 @@ public class RobotPoseSource {
-                pose.getRotation().getY(),
-                pose.getRotation().getZ()
-        };
-    }
-
     public void update() {
         robotPoseSourceIO.updateInputs(inputs);
         Logger.processInputs("Cameras/" + name, inputs);
-        cachedPose = getUnCachedRobotPose();
-        if (!inputs.hasResult || inputs.averageDistanceFromTags == 0 || cachedPose == null)
+
+        robotPose = calculateBestRobotPose(inputs.averageDistanceFromBestTag);
+
+        if (!inputs.hasResult || inputs.averageDistanceFromAllTags == 0 || robotPose == null)
             Logger.recordOutput("Poses/Robot/" + name + "Pose", RobotPoseSourceConstants.EMPTY_POSE_LIST);
         else
-            Logger.recordOutput("Poses/Robot/" + name + "Pose", cachedPose);
+            Logger.recordOutput("Poses/Robot/" + name + "Pose", robotPose);
     }
 
-    public int getVisibleTags() {
-        return inputs.visibleTags;
+    public boolean hasNewResult() {
+        return (inputs.hasResult && inputs.averageDistanceFromAllTags != 0) && isNewTimestamp();
     }
 
-    public double getAverageDistanceFromTags() {
-        return inputs.averageDistanceFromTags;
+    /**
+     * @return the current estimated robot position
+     */
+    public Pose2d getCurrentRobotPose() {
+        return robotPose;
     }
 
-    public boolean hasNewResult() {
-        return (inputs.hasResult && inputs.averageDistanceFromTags != 0) && isNewTimestamp();
+    public String getName() {
+        return name;
     }
 
-    public Pose2d getRobotPose() {
-        return cachedPose;
+    public double getAverageDistanceFromTags() {
+        return inputs.averageDistanceFromAllTags;
     }
 
-    public String getName() {
-        return name;
+    public int getNumberOfVisibleTags() {
+        return inputs.visibleTagIDs.length;
     }
 
     public double getLastResultTimestamp() {
         return inputs.lastResultTimestamp;
     }
 
-    private Pose2d getUnCachedRobotPose() {
-        final Pose3d cameraPose = doubleArrayToPose3d(inputs.cameraPose);
+    /**
+     * If the robot is close enough to the tag, it gets the estimated solve PNP pose.
+     * If it's too far, it assumes the robot's heading and calculates its position from there.
+     *
+     * @param averageDistanceFromBestTag the average of the distance from the best visible tag
+     * @return the robot's pose
+     */
+    private Pose2d calculateBestRobotPose(double averageDistanceFromBestTag) {
+        if (averageDistanceFromBestTag < RobotPoseSourceConstants.MAXIMUM_DISTANCE_FROM_TAG_FOR_PNP_METERS)
+            return cameraPoseToRobotPose(inputs.solvePNPPose);
+        return calculateAssumedRobotHeadingPose().toPose2d();
+    }
+
+    /**
+     * Calculates the robot's pose by assuming its heading.
+     *
+     * @return the robot's pose
+     */
+    private Pose3d calculateAssumedRobotHeadingPose() {
+        final Rotation2d robotHeading = RobotContainer.SWERVE.getHeading();
```

Comment:
This should sample the heading at the result's timestamp from the 6328 pose estimator's buffer